### PR TITLE
Xattr

### PR DESCRIPTION
**Pull request by kahing**
no tests written yet in this repo but I've tried them in goofys side and they seem to work. Want to get some input on interfaces and SetXattrOp.Flags handling

### CODE REVIEW

File: fuseops/ops.go [hunk 1]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
 	// Set by the file system: the target of the symlink.
 	Target string
 }
+
+////////////////////////////////////////////////////////////////////////
+// eXtended attributes
+////////////////////////////////////////////////////////////////////////
+
+// Remove an extended attribute
```
Comment (jacobsa) [inline fuseops/ops.go]: Please add comments that give some context about what this is and when it's sent. See `OpenFileOp` for example.

File: fuseops/ops.go [hunk 2]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+type RemoveXattrOp struct {
+	// The inode that we are reading
+	Inode InodeID
+
+	// The name of the extended attribute
+	Name string
+}
+
+// Get an extended attribute
```
Comment (jacobsa) [inline fuseops/ops.go]: Context comments here too.

File: fuseops/ops.go [hunk 3]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
 	Target string
 }
+
+////////////////////////////////////////////////////////////////////////
+// eXtended attributes
+////////////////////////////////////////////////////////////////////////
+
+// Remove an extended attribute
+type RemoveXattrOp struct {
+	// The inode that we are reading
```
Comment (jacobsa) [inline fuseops/ops.go]: Full stops at the end of each description please.

File: fuseops/ops.go [hunk 4]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+	Inode InodeID
+
+	// The name of the extended attribute
+	Name string
+}
+
+// Get an extended attribute
+type GetXattrOp struct {
+	// The inode that we are reading
```
Comment (jacobsa) [inline fuseops/ops.go]: s/we are reading/whose extended attribute we are reading/

(Here and below.)

File: fuseops/ops.go [hunk 5]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+	Inode InodeID
+
+	// The name of the extended attribute
```
Comment (jacobsa) [inline fuseops/ops.go]: Add comments: what should the FS do if the attribute doesn't exist?

File: fuseops/ops.go [hunk 6]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+	Name string
+
+	// The destination buffer.  If the size is too small for the
+	// value, the ERANGE error should be sent.
+	Dst []byte
+
+	// Set by the file system: the number of bytes read into Dst, or
+	// the number of bytes that would have been read into Dst if Dst was
```
Comment (jacobsa) [inline fuseops/ops.go]: How does this interact with the requirement to return `ERANGE` above?

File: fuseops/ops.go [hunk 7]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+	// big enough
+	BytesRead int
+}
+
+type ListXattrOp struct {
```
Comment (jacobsa) [inline fuseops/ops.go]: Overview documentation here too.

File: fuseops/ops.go [hunk 8]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+	// one for each xattr
+	Dst []byte
+
+	// Set by the file system: the number of bytes read into Dst, or
+	// the number of bytes that would have been read into Dst if Dst was
+	// big enough
+	BytesRead int
+}
+
+type SetXattrOp struct {
```
Comment (jacobsa) [inline fuseops/ops.go]: Overview documentation here too.

File: fuseops/ops.go [hunk 9]
```diff
@@ -756,3 +756,66 @@ type ReadSymlinkOp struct {
...
+	// The inode that we are changing
+	Inode InodeID
+
+	// The name of the extended attribute
+	Name string
+
+	// The data to for the extened attribute.
+	Data []byte
```
Comment (jacobsa) [inline fuseops/ops.go]: The man page for `setxattr` seems to call this "value". Maybe use that word instead?

**Comment by kahing** [comment]
addressed review comments

File: samples/memfs/memfs_test.go [hunk 10]
```diff
@@ -1611,6 +1613,83 @@ func (t *MemFSTest) RenameNonExistentFile() {
...
+
+func (t *MemFSTest) SetXAttr() {
+	var err error
+
+	// Create a file
+	filePath := path.Join(t.Dir, "foo")
+	err = ioutil.WriteFile(filePath, []byte("taco"), 0600)
+	AssertEq(nil, err)
+
+	err = xattr.Setxattr(filePath, "foo", []byte("bar"), 0x2)
```
Comment (jacobsa) [inline samples/memfs/memfs_test.go]: These magic constants (`0x2`, `0x1`) are confusing and hard to maintain. Please either upstream symbolic names for them into the `go-xattr` package, or add your own xattr package in this repo (probably in an `internal` directory) that has symbolic names for them.

Comment (kahing) [inline samples/memfs/memfs_test.go]: opened upstream PR for the constants

**Comment by jacobsa** [comment]
Thanks. Please ping this when ivaxer/go-xattr#1 is resolved.

**Comment by kahing** [comment]
upstream doesn't seem active, any opinion on having the constants in this project?

**Comment by jacobsa** [comment]
This project really doesn't seem like the place to put them. I would recommend either adding the entire functionality of that project to an `internal/` project here (complete with constants), or forking that package and fixing the issue in your fork.

**Comment by kahing** [comment]
maintaining a fork for 2 constants seems silly, an argument for putting them here is that any fuse filesystems that want to implement xattrs will want to use those constants too. In a way it's not that different from other fuse error codes that you already export.

**Comment by jacobsa** [comment]
Forking an unmaintained project for the sake of making a reasonable change to it is a totally sane thing to do. The impetus for the change doesn't matter; the point is that it's unmaintained.

Stuffing two constants here that are relevant only for an entirely different project doesn't make sense. Please either fix the upstream project (by upstreaming or forking), or pull the entirety of its functionality here, and give it a sane API.

**Comment by kahing** [comment]
amended last commit to switch to my fork of go-xattr

**Comment by jacobsa** [comment]
Thanks!

**Pull request merged**